ENH: improve rayleigh accuracy for large x.

maniteja123 opened the pull request:
Improve the accuracy of the rayleigh distribution for large x, by replacing exp(x)-1 with expm1(x) and log(1+x) with log1p(x). closes #4034

larsmans commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3542,10 +3542,16 @@ def _pdf(self, r):
 
     def _cdf(self, r):
-        return 1 - exp(-0.5 * r**2)
+        return -special.expm1(-r * r / 2.0)
```
I agree with using `expm1` here, but not the rest of the change.

* `-r` allocates an array;
* `r / 2.0` allocates an array;
* multiplying these allocates an array.

By contrast, `-0.5 * r ** 2` allocates only two arrays. Also, division has to check for a zero second argument, so multiplication is to be preferred if the divisor is constant.

maniteja123 replied:
@larsmans , Oh I didn't know about the memory implications. Thanks for information. Shall I revert back to the original formula.

maniteja123 replied:
@larsmans , Oh I didn't know about the memory implications. Thanks for information. Shall I revert back to the original formula.

larsmans replied:
Yes, please! Your `_sf` also uses the less efficient idiom, it seems.

maniteja123 commented:
@larsmans. Sure, will do it now. Just needed some advice. The present branch is some commits behind the master. Do I need to rebase before committing. Last time I did it, I accidentally closed the PR. It would be great if you could clarify this.

pv commented:
You don't need to rebase, and it's best if you don't. See more "best practice" details in these instructions: http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html

maniteja123 commented:
@pv Thanks :). I have gone through the documentation. I will just push the changes soon.

maniteja123 commented:
I have a small doubt, right now ``git br -v`` shows this:

    issue4034 c88c142 [ahead 8, behind 3] ENH: Updated tests for lomax and rayleigh accuracy 
    Take care of floating point comparison in tests.closes #4034
 
I don't want to accidentally close this PR again. I have squashed all the commits and made into a single commit. If I do ``git push origin issue4034``, it rejects it saying 

    To git@github.com:maniteja123/scipy.git
    ! [rejected]        issue4034 -> issue4034 (non-fast-forward)
    error: failed to push some refs to 'git@github.com:maniteja123/scipy.git'
    hint: Updates were rejected because the tip of your current branch is behind
    hint: its remote counterpart. Integrate the remote changes (e.g.
    hint: 'git pull ...') before pushing again.
    hint: See the 'Note about fast-forwards' in 'git push --help' for details.

I am not sure whether doing ``git push origin issue4034 -f`` will solve the problem. Sorry, I am not that accustomed to git. Any help would be appreciated.

rgommers commented:
If you see for your local ``issue4034`` branch that it has your single commit on top of ``master``, then you can do

    git push origin issue4034 -f    # origin is your own scipy github repo

That should not close this PR.

**However**: I don't understand why your ``git br -v`` output says "8 ahead" when you say you only have a single commit. Can you paste the output of ``git log`` up till and including the first commit that is already in scipy master?

maniteja123 commented:
@rgommers Thanks :). Sorry for my doubts. The [remote](https://github.com/maniteja123/scipy/tree/issue4034) branch is 3 commits ahead, 62 commits behind scipy:master. I now force pushed from my local.

I have done 3 commits before, but now I squashed them to one single commit. This is the output of the ``git log``. https://www.dropbox.com/s/9akwdjycoo010ha/a.txt?dl=0

ev-br commented:
The branch comparison, https://github.com/maniteja123/scipy/compare/issue4034, seems clean enough with a single commit compared to scipy master...

maniteja123 commented:
@ev-br Oh, I get it now, with scipy master as base, my branch is ahead by only 1 commit. So isn't it a problem how many commits scipy master is ahead than my branch or master ?

rgommers commented:
That's correct. But it's easiest if you keep your own master branch in your local repo in sync with scipy master, then you don't have to worry which master is where.

maniteja123 commented:
@rgommers @ev-br Thank you for clarification. I will keep my master in sync with scipy master, but by the time I push my feature branch, the scipy master generally goes ahead by many commits. So, I was confused whether I need to update my feature branch to sync with the present state of scipy master before ``git push``. But I got it now. :)

rgommers commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2171,12 +2171,17 @@ def test_infinite_input():
     # regression test for gh-4033
     p = stats.lomax.ppf(stats.lomax.cdf(1e-100,1),1)
-    assert_equal(p, 1e-100)
+    assert_almost_equal(p, 1e-100, decimal=15)
```
This doesn't work, because ``assert_almost_equal`` does absolute comparison with 1e-15 precision here:

    assert_almost_equal(2e-100, 1e-100, decimal=15)

I suggest instead ``assert_allclose(p, 1e-100)``. That uses relative comparison with rtol=1e-7

rgommers commented:
The changes to ``rayleigh`` and test for it look good now.

maniteja123 commented:
@rgommers Sorry, I was following the suggestion you gave in issue #4423, where you mentioned floating point comparisons. I think I should take time to look into the way ``assert`` statements work. I will change it to ``assert_allclose(p, 1e-100)`` in ``rayleigh`` and ``truncexpon`` #4425 also.

rgommers commented:
Thanks @maniteja123, merged.

rgommers commented:
Playing around a bit with the different ``assert_`` functions will be useful, you'll find differences in behavior between relative and absolute tolerances that are important to understand.
